lib/discoverable.coffee

```diff
@@ -0,0 +1,226 @@
+				service = ''
+				tags = []
+				if components.length == 3
+					service = "_#{components[0]}._sub."
+					components.shift()
+				service += "_#{components[0]}._#{components[1]}"
+
+				fs.readFileAsync("#{registryPath}#{path}/tags.json", { encoding: 'utf8' })
+				.then (data) ->
+					json = JSON.parse(data)
+					if (!_.isArray(json))
+						throw new Error()
+
+					tags = json
+				.catch (err) ->
+					# If the tag file didn't exist, we silently fail.
+					if (err.code != 'ENOENT')
+						throw new Error("tags.json for #{service} service defintion is incorrect")
+				.then ->
+					services.push({ service: service, tags: tags })
+		.return(services)
+
+# Set the services function as a memoized one.
+services = _.memoize(retrieveServices)
+
+###
+# @summary Sets the path which will be examined for service definitions.
+# @function
+# @public
+#
+# @description
+# Should no parameter be passed, or this method not called, then the default
+# path is the 'services' directory that exists within the module's directory
+# hierarchy.
+#
+# @param {String} path - New path to use as the service registry.
+#
+# @example
+# discoverableServices.setRegistryPath("/home/heds/discoverable_services")
+###
+exports.setRegistryPath = (path) ->
+	if !path?
+		path = "#{__dirname}/../services"
+
+	if !_.isString(path)
+		throw new Error('path parameter must be a path string')
+
+	registryPath = path
+	services.cache.clear()
+
+###
+# @summary Enumerates all currently registered services available for discovery.
+# @function
+# @public
+#
+# @description
+# This function allows promise style if the callback is omitted.
+#
+# @param {Function} callback - callback (error, services)
+#
+# @example
+# discoverableServices.enumerateServices (error, services) ->
+#   throw error if error?
+#   # services is an array of service objects holding type/subtype and any tagnames associated with them
+#   console.log(services)
+###
+exports.enumerateServices = (callback) ->
+	if callback? and !_.isFunction(callback)
+		throw new Error('callback parameter must be a function')
+
+	services()
+	.asCallback(callback)
+
+###
+# @summary Listens for all locally published services, returning information on them after a period of time.
+# @function
+# @public
+#
+# @description
+# This function allows promise style if the callback is omitted. Should the timeout value be missing
+# then a default timeout of 2000ms is used.
+#
+# @param {Array} services - A string array of service names or tags
+# @param {Number} timeout - A timeout in milliseconds before results are returned. Defaults to 2000ms
+# @param {Function} callback - callback (error, services)
+#
+# @example
+# discoverableServices.findServices([ '_resin-device._sub._ssh._tcp' ], 5000, (error, services) ->
+#	throw error if error?
+#   # services is an array of every service that conformed to the specified search parameters
+#   console.log(services)
+###
+exports.findServices = (services, timeout, callback) ->
```

Comment:
This can be

```coffeescript
exports.findServices = Promise.method (services, timeout) ->
```

So any returned value in the body of the function will resolve the promise and any error will reject it.